resolver-dns/src/main/java/io/netty/resolver/dns/DnsQueryIdSpace.java

```diff
@@ -0,0 +1,114 @@
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.resolver.dns;
+
+import io.netty.util.internal.MathUtil;
+import io.netty.util.internal.PlatformDependent;
+
+import java.util.Random;
+
+/**
+ * Special data-structure that will allow to retrieve the next query id to use, while still guarantee some sort
+ * of randomness.
+ * The query id will be between 0 (inclusive) and 65535 (inclusive) as defined by the RFC.
+ */
+final class DnsQueryIdSpace {
+    private static final int MAX_ID = 65535;
+
+    // Holds all possible ids which are stored as unsigned shorts
+    private final short[] ids = new short[MAX_ID + 1];
+    private final int mask = ids.length - 1;
+    private int head;
+    private int tail;
+    private int count;
+
+    DnsQueryIdSpace() {
+        assert ids.length == MathUtil.findNextPositivePowerOfTwo(MAX_ID);
+        for (int v = 0; v <= MAX_ID; v++) {
+            pushId(v);
+        }
+    }
+
+    /**
+     * Returns the next ID to use for a query or {@code -1} if there is none left to use.
+     *
+     * @return  next id to use.
+     */
+    int nextId() {
+        assert count >= 0;
+        if (count == 0) {
+            return -1;
+        }
+        short id = ids[head];
+        head = (head + 1) & mask;
+        count--;
+
+        return id & 0xFFFF;
+    }
+
+    /**
+     * Push back the id, so it can be used again for the next query.
+     *
+     * @param id    the id.
+     */
+    void pushId(int id) {
+        if (count == ids.length) {
+            throw new IllegalStateException("overflow");
+        }
+
+        assert id <= MAX_ID && id >= 0;
+        if (tail == head) {
+            ids[tail] = (short) id;
+        } else {
+            // Let's ensure our ids will be returned in a random fashion by not always add the id to the tail.
+            final int idx;
+            Random random = PlatformDependent.threadLocalRandom();
+            if (tail < head) {
+                if (tail != 0 && random.nextBoolean()) {
+                    idx = random.nextInt(tail);
+                } else {
+                    idx = random.nextInt(ids.length - head) + head;
+                }
+            } else {
+                idx = random.nextInt(tail - head) + head;
+            }
+            short otherId = ids[idx];
+            ids[idx] = (short) id;
+            ids[tail] = otherId;
+            assert otherId != (short) id;
+        }
+        tail = (tail + 1) & mask;
+        count++;
+    }
```

Comment:
I think you can make this substantially simpler by just making the insertion position random and always pulling from the tail. That leaves the only state left being `ids` and `count`.
```suggestion
    void pushId(int id) {
        if (count == ids.length) {
            throw new IllegalStateException("overflow");
        }

        assert id <= MAX_ID && id >= 0;
        // pick a slot for our index, and whatever was in that slot before will get moved to the tail.
        Random random = PlatformDependent.threadLocalRandom();
        int insertionPosition = random.nextInt(count + 1);
        ids[count + 1] = ids[insertionPosition];
        ids[insertionPosition] = (short) id;
        count++;
    }
```

Reply:
And if you want to make it more obviously random we can move this to the `nextId()` method where it simply selects a random position up to `count`, used that as the result, and moves the tail into the hole that was made.

Reply:
Thanks a lot! This is a great idea <3 . I adjusted the code to follow your idea